lib/aws-s3.js

```diff
@@ -403,19 +403,28 @@ module.exports = class AWSAttachmentsService extends require("./basic") {
    * @param {import('@sap/cds').Request} req - The request object
    */
   async attachDeletionData(req) {
-    const attachments = cds.model.definitions[req?.target?.name + ".attachments"]
-    if (attachments) {
-      const diffData = await req.diff()
-      let deletedAttachments = []
-      diffData.attachments?.filter((object) => {
-        return object._op === "delete"
-      })
-        .map((attachment) => {
-          deletedAttachments.push(attachment.ID)
+    const attachmentCompositions = Object.keys(req?.target?.associations).filter(assoc => req?.target?.associations[assoc]._target['@_is_media_data']);
+    if (attachmentCompositions.length) {
+      const diffData = await req.diff();
+      const queries = [];
+      for (const attachmentsComp of attachmentCompositions) {
+        let deletedAttachments = []
+        diffData[attachmentsComp]?.forEach(object => {
+          if (object._op === "delete") {
+            deletedAttachments.push(object.ID)
+          }
         })
-
-      if (deletedAttachments.length > 0) {
-        let attachmentsToDelete = await SELECT.from(attachments).columns("url").where({ ID: { in: [...deletedAttachments] } })
+        if (deletedAttachments.length) {
+          queries.push(
+            SELECT.from(req.target.associations[attachmentsComp]._target).columns("url").where({ ID: { in: [...deletedAttachments] } })
+          )
+        }
+      }
+      if (queries.length > 0) {
+        let attachmentsToDelete = []
+        for (const toDeleteAttachments of await Promise.all(queries)) {
+          attachmentsToDelete = attachmentsToDelete.concat(toDeleteAttachments);
+        }
```

Comment:
```suggestion
        const attachmentsToDelete = (await Promise.all(queries))
          .reduce((all, current) => all.concat(current), []);
```

or

```suggestion
        const attachmentsToDelete = (await Promise.all(queries))
           .reduce((all, current) => [...all, ...current], []);
```

Reply:
```suggestion
        const attachmentsToDelete = (await Promise.all(queries)).flat()
```

would that work, I think > Node11 supports flatten arrays